File: MatlabLibrary/spline_inverse/steepestdescent.m
function [x,dfdx,alpha]=steepestdescent(phi,dphidx,xinit,mingradnorm,wolfeparams)
% Algorithm: Steepest Descent
% Author: Burak Erem
% --------------------------------------
% INPUT: phi = objective function
%        dphidx = gradient of objective function
%        xinit = initialization of algorithm step sequence
%        mingradnorm = smallest 2-norm of dphidx before stopping
% --------------------------------------
% OUTPUT: x = matrix whose columns are the sequence of x points
%         dfdx = matrix whose columns are the gradients at x
%         alpha = array of the step size parameters used

% Initialization:
x=xinit;
f=phi(x);
dfdx=dphidx(x);
p=-dfdx;
k=1;
alpha(1)=1;

alphamin=1e-3;alpharate=1.5;alphamax=10;c1=1e-2;c2=0.6;
if(isfield(wolfeparams,'alphamin')),alphamin=wolfeparams.alphamin;end;
if(isfield(wolfeparams,'alphamax')),alphamax=wolfeparams.alphamax;end;
if(isfield(wolfeparams,'alpharate')),alpharate=wolfeparams.alpharate;end;
if(isfield(wolfeparams,'c1')),c1=wolfeparams.c1;end;
if(isfield(wolfeparams,'c2')),c2=wolfeparams.c2;end;

costs(k)=phi(x(:,k));
costreduction=inf;

fprintf('Init objfun=%g\n',costs(k))

% The main loop
while(costreduction>mingradnorm)

%     Line Search (satisfies strong Wolfe conditions)
    alpha(k)=linesearch(phi,dphidx,x(:,k),p(:,k),c1,c2,alphamin,alpharate,alphamax);

%     Set next sequence step x_(k+1)
    x(:,k+1)=x(:,k)+alpha(k)*p(:,k);

%     Evaluate gradient(f_(k+1))
    dfdx(:,k+1)=dphidx(x(:,k+1));

%     Set search direction p_(k+1)
    p(:,k+1)=-dfdx(:,k+1);
    
    costs(k+1)=phi(x(:,k+1));
    costreduction=costs(k)-costs(k+1);
    
%     Report progress on the current iteration
    fprintf('k=%i\t\tobjfun=%g\t\tcostreduction=%g\n',k,costs(k+1),costreduction)

%     Increment k
    k=k+1;

end

File: PythonLibrary/spline_inverse/steepestdescent.py
import numpy as np
from linesearch import linesearch


def steepestdescent(phi, dphidx, xinit, mingradnorm, alphamax=10):
    # Algorithm: Steepest Descent
    # Author: Burak Erem
    # --------------------------------------
    # INPUT: phi = objective function
    #        dphidx = gradient of objective function
    #        xinit = initialization of algorithm step sequence
    #        mingradnorm = smallest 2-norm of dphidx before stopping
    # --------------------------------------
    # OUTPUT: x = matrix whose columns are the sequence of x points
    #         dfdx = matrix whose columns are the gradients at x
    #         alpha = array of the step size parameters used
    #

    # Initialization:
    k = 0
    x = []
    x.append(xinit)
    f = phi(x[k])
    dfdx = []
    dfdx.append(dphidx(x[k]))
    p = []
    p.append(-dfdx[k])

    alpha = []

    costs = []
    costs.append(phi(x[k]))
    cost_reduction = np.inf

    print('Init obj fun={0}'.format(costs[k]))

    # The main loop
    while (cost_reduction > mingradnorm):
        # Line Search (satisfies strong Wolfe conditions)
        alpha.append(linesearch(phi, dphidx, x[k], p[k], alphamax))

        # Set next sequence step x_(k+1)
        x.append(x[k] + alpha[k] * p[k])

        # Evaluate gradient(f_(k+1))
        dfdx.append(dphidx(x[k + 1]))

        # Set search direction p_(k+1)
        p.append(-dfdx[k + 1])
        costs.append(phi(x[k + 1]))
        cost_reduction = costs[k] - costs[k + 1]

        # Report progress on the current iteration
        print('k={0}    objfun={1}      costreduction={2}'.format(k, costs[k + 1], cost_reduction))

        # Increment k
        k = k + 1

    return x, dfdx, alpha


if __name__ == '__main__':
    N = 2
    phi = lambda x: np.sum(x ** 2)
    dphidx = lambda x: 2 * np.ones((N, 1))
    xinit = np.ones((N, 1))
    alphamax = 0.1
    mingradnorm = 1e-15

    x, dfdx, alpha = steepestdescent(phi, dphidx, xinit, mingradnorm, alphamax)
    print(x[-1])


